### PR TITLE
[forth] add test ruling out eager expansion of custom arguments.

### LINKED ISSUE #1221: [forth] Add test ruling out eager expansion of custom arguments
I see several implementations which eagerly expand custom arguments such that their dictionary of names contains only primitives. This is a bad idea. Consider the program:

```forth
  : attack-a 0 drop ;
  : attack-b attack-a attack-a ;
  : attack-c attack-b attack-b ;
  // ...
  : attack-z attack-y attack-y ;
```

Given eager argument expansion, `definitions["attack-z"].len() == 2.pow(26)` (26 Mb), which is an exorbitant amount of memory for something defined with two instructions. It would be trivial to extend the attack further to cause an OOM on any arbitrary machine. 

Challenge: we don't want to force students into using any particular data structure for storing their instructions map. 

One approach satisfying the challenge might be to write a custom global allocator for the test module which keeps track of how much memory has been allocated, and provides some interface to query that value at runtime. We could then write a test like the above attack, and check the allocation quantity. As long as it's under, say, 1 Mb, then the test should pass. See [`std::alloc`](https://doc.rust-lang.org/std/alloc/index.html).

### PR DESCRIPTION
**Pull request by coriolinus**
Closes #1221.

Implements the approach suggested in that issue.

### CODE REVIEW

**Comment by coriolinus** [comment]
Note that the current example solution suffers from the attack this PR is meant to ban. This means that we need to update the example.

I'd ideally use a [student solution](https://exercism.io/mentor/solutions/b6f9f69b03df4b889c9930960cb4a358?iteration_idx=8) which is both compact and readable, but I'm not happy with the ethics of just grabbing it without permission, even with attribution. I think that approach is only viable if the student approves, and I don't have a good way to contact them. I don't think that mentor notes are the appropriate venue for such a request.

I'm happy to contribute [my own solution](https://github.com/coriolinus/exercism-rs/blob/master/forth/src/lib.rs), but that solution has not been reviewed at all and requires an external library. It's very readable (IMO), but I got a little architecture-happy with it.

**Comment by iHiD** [comment]
We don't need permission. Everything a student uploads is licensed to us to use as we want. That's made very explicit during the signup process. 

Once it's done though I can let them know :)

**Comment by coriolinus** [comment]
Thanks @iHiD, I think they'd appreciate knowing that once this is merged.

File: exercises/practice/forth/tests/alloc-attack.rs [hunk 1]
```diff
@@ -0,0 +1,96 @@
+//! Certain implementations of Forth naively and eagerly expand sub-definitions
```
Comment (efx) [inline exercises/practice/forth/tests/alloc-attack.rs:1]: Love the added context here!

**Review by efx — APPROVED** [review]

**Pull request merged**